csharp/src/Ice/TcpAcceptor.cs

```diff
@@ -16,12 +19,33 @@ internal class TcpAcceptor : IAcceptor
         private readonly IConnectionManager _manager;
         private readonly Socket _socket;
         private readonly IPEndPoint _addr;
+        private const byte TlsHandshakeRecord = 0x16;
 
         public async ValueTask<Connection> AcceptAsync()
         {
             Socket fd = await _socket.AcceptAsync().ConfigureAwait(false);
 
-            ITransceiver transceiver = ((TcpEndpoint)Endpoint).CreateTransceiver(fd, _adapter.Name);
+            bool secure = Endpoint.IsAlwaysSecure || !_adapter.AcceptNonSecure;
+
+            if (_adapter.Protocol == Protocol.Ice2 && _adapter.AcceptNonSecure)
+            {
+                Debug.Assert(_adapter.Communicator.ConnectTimeout != TimeSpan.Zero);
+                // TODO: we are using reusing ConnectTimeout here so that peeking cannot block forever.
+                // However, this means that it's possible to end up waiting 2 * ConnectTime if reading the
+                // first byte is slow and then the actual connection initialization is also slow.
+
+                using var source = new CancellationTokenSource(_adapter.Communicator.ConnectTimeout);
```

Comment:
Calling `InitializeAsync` from inside `AcceptAsync` with `Task.Run()` seems to cause some issues (without the token changes to InitializeAsync).

```
*** [27/52] Running csharp/Ice/metrics tests ***
[ running client/server test - 11/12/20 12:03:45 ]
testing group by none...ok
testing group by id...ok
testing connection metrics...

System.Threading.Tasks.TaskCanceledException: A task was canceled.
```

I suspect because we're not waiting for the ConnectionFactory to finish [getting ready.](https://github.com/zeroc-ice/ice/blob/master/csharp/src/Ice/ConnectionFactory.cs#L615-L620). 

What if `AcceptAsync` returns a connection and an un-started task which it can start when it's ready? I tested it and it works.

````
(connection, initializeTask) = await _acceptor.AcceptAsync();

// We don't wait for the connection to be activated. This could take a while for some transports
// such as TLS based transports where the handshake requires few round trips between the client
// and server.
initializeTask.Start();
...

Reply:
This issue is actually due to the fact that moving the `InitializeAsync` call into the AcceptAsync causes it to run on an unexpected scheduler. The plan is to leave this for now, and refactor SSLTransceiver and TCPTransceiver into one class and have to do the peeking.